edn_format/edn_parse.py

```diff
@@ -98,6 +108,32 @@ def p_map(p):
     p[0] = ImmutableDict(dict([terms[i:i + 2] for i in range(0, len(terms), 2)]))
 
 
+def p_discarded_expression(p):
```

Comment:
Seems like a lot of work for discard expression. I feel bad asking this, because you have put in so much effort, have you compared this to other implementations - do they also have several rules to handle discarding?

Reply:
I’d love to remove all that code and replace it with a simpler implementation :smile: 

I’ll check if we can do better but with this grammar-based parsing I don’t think we can: We have three parsing stages: tokenizing, lexing and parsing.
* We can’t handle it during tokenizing/lexing because discarding works on whole expressions, not single tokens. For example, `#_ [] 2` should be read as `2`, not `] 2`; i.e. we must "see" the list before discarding it.
* We can’t do it after parsing because discarding affects the way we parse data structures: `{:a #_ #_ 1 2 :b}` should be read as `{:a :b}`, not `{:a #_, #_ 1, 2 :b}`. `[#_ #_ 1 2 3]` can’t be parsed as `[#_(#_(1)) 2 3]`; it should be parsed as `[#_ 2 3]` then `[3]`.

Reply:
Also, something I’ve learnt maintaining open-source projects: don’t judge a PR by the amount of work the contributor put in.

Quoting [Mike McQuaid](https://mikemcquaid.com/2018/08/14/the-open-source-contributor-funnel-why-people-dont-contribute-to-your-open-source-project/) : 
> No “guilty merges”. Don’t merge work from contributors you know isn’t good enough just because you feel bad about the work they’ve put in.